src/core/Elsa.Core/Extensions/ElsaServiceCollectionExtensions.cs

```diff
@@ -75,9 +97,29 @@ public static class ElsaServiceCollectionExtensions
 
         public static IServiceCollection StartWorkflow<T>(this IServiceCollection services) where T : class, IWorkflow => services.AddHostedService<StartWorkflow<T>>();
 
-        private static IServiceCollection AddMediatR(this ElsaOptions options) => options.Services.AddMediatR(mediatr => mediatr.AsScoped(), typeof(IActivity));
+        /// <summary>
+        /// Add all workflows (<see cref="IWorkflow"/>) that are in the assembly and have no constructor with parameters
```

Comment:
We have to be able to add `IWorkflow` implementations that do receive constructor arguments.

Reply:
I'd like it that way too, but it doesn't work for one example. In one example Instant is passed as a parameter. This is not registered in the DI, which gives an error.

If you don't mind, I would change the example to also add workflows with parameters.

Reply:
In that case, the user should provide their own workflow factory method (as is also shown in that example).
But I see why it wouldn't work with `IWorkflow` types now being part of `ElsaOptions`, because all you do there is register workflow types. Perhaps we ought to consider workflows not to be part of ElsaOptions, but regular services instead (as is the situation before the changes introduced here).

Reply:
We still have the problem. In `ProgrammaticWorkflowProvider` I need the instance of the workflow. All existing workflows are instantiated.  So here we have the problem again, as soon as a constructor has a parameter, which is not in the DI.

Reply:
Not if the user registers their workflow like this:

```csharp
services.AddWorkflow(new SomeWorkflow("Foo"))
```

The above registers `SomeWorkflow` as a singleton.
In fact, we could simply add workflow instances to `ElsaOptions` instead of just their types. Or at the very least, we could register workflow factory methods, allowing the caller to provide a workflow instance as they see fit.

Reply:
That would work, but then concrete instantiations would have to be made before calling `public static IServiceCollection AddWorkflow(this IServiceCollection services, Assembly assembly)`.

I have not yet dealt with factory methods. I would have to think about that.

Reply:
Not necessarily; the `IServiceCollection` API allows you to specify a factory method which gets invoked whenever an object instance is requested (and only executed once when life time scope is Singleton).

Anyway, see what you can do. If you want we can always revisit this later after we merged in this PR.

Reply:
Good idea. I was thinking too complicated.

Reply:
Is added it now. Is this what you had in mind?

Unfortunately I don't know how to resolve conflicts with Github and a pull request.

Reply:
Almost. Right now you are allowing the user to add an instance of a workflow, but then you only register its type and discard the object itself.

But that doesn't have to block this PR as far as I'm concerned - we can work on this after merge.
I'll work on the merge conflicts, but let me know if you want me to show you.

Reply:
Sorry. I was a little careless there. When I wrote that, I was already thinking about the next method.

I have corrected it now.

Reply:
No worries, that looks good. Although I am still not sure what the point is of registering the type or even the instance with the service container *and also* with `ElsaOptions`. I think we only need to do one *or* the other.

If this is OK with you, I'd like to get your PR merged in regardless. and make changes afterwards.

Reply:
I did it just in case someone wants to add a concrete instance. 

With `workflows.Add((IWorkflow)ActivatorUtilities.CreateInstance(serviceProvider, workflow));` in `ProgrammaticWorkflowProvider` this instance should be retrieved.

However, if you would like it to be different, feel free to change it.